docs/developer/best-practices/stability.asciidoc

```diff
@@ -43,11 +43,30 @@ dependency list!
 [discrete]
 === Test coverage
 
-* Does the feature have sufficient unit test coverage? (does it handle
-storeinSessions?)
-* Does the feature have sufficient Functional UI test coverage?
-* Does the feature have sufficient Rest API coverage test coverage?
-* Does the feature have sufficient Integration test coverage?
+Testing UI code is hard. Here are our heuristics for maintaining Kibana's quality with tests.
+
+**Priority 1:** Our first priority is to minimize the total number of poor user experiences caused by bugs, and to prevent bugs that can cause harm to the user. "Harm" includes direct problems like data loss and data entering a bad state, as well as indirect problems like making a poor business decision based on misinformation presented by the UI.
+
+**Priority 2:** Our second priority is to define all correct UX behavior and guard against any UX regressions.
+
+**We are constrained by time.** We don’t have enough time to test all possible input and output permutations of the UX. Just think of the different ways a form can be interacted with and respond to input. We must spend our time prudently to address the above priorities.
+
+With each feature we ship, we have a **hard requirement** for test coverage of common library code, popular UIs, critical code (e.g. payload de/serialization, server-side APIs), and critical features (e.g. visualization, security). We have a **soft requirement** for test coverage of everything else.
```

Comment:
@cjcenizal - thanks for pulling this PR together, I think it's great we add more information on what is expected for test coverage.

I know we had a conversation about framing it like this. How we know it's impossible to cover every permutation, and we also know that sometimes exceptions are made based on business priorities. I wonder if mentioning this explicitly will help or hurt though. I have a very hard time thinking of a situation where a developer added _too many_ tests. I also can't think of a situation where a business critical feature was pushed off because of too little tests and we retrospectively thought that was a _bad_ thing. Almost every case I can think of is the opposite. I wish we didn't rush this feature out, I wish we added more tests. I'm so glad we didn't rush that feature out, it's now solid quality, no bug reports!

Is there a need to explicitly mention these corner cases?  I don't think product, team and tech leads will feel like they can't make exceptions just because we don't have it written down here.  I worry it could be abused and end up as a poor excuse for less test coverage.

I tried my hand at rewording this section with the goal of helping engineers focus their energy on the most important areas, without trying to create any hard rules.  What do you think??

----

Testing UI code is hard. 100% test coverage is ideal, but hard to measure, and time is limited, so how do you know you if the tests you wrote offer sufficient coverage to ensure code quality and adhere by our Developer Principal of [extensive automatic testing through ci](https://github.com/elastic/engineering/blob/master/kibana_dev_principles.md#automate-tests-through-ci)? We are working on incorporating unit code coverage statistics on every PR, but it ultimately comes down to a judgement call.  Here are some guidelines to help you ensure sufficient test coverage.

Every PR should be accompanied by tests. 

Check the before and after unit test coverage metrics and ensure the number didn't go down.

Pay special attention to code that if implemented incorrectly can cause harm to the user. "Harm" includes direct problems like data loss and data entering a bad state, as well as indirect problems like making a poor business decision based on misinformation presented by the UI. For example, state migrations and security permissions are very important areas to cover.

Pay special attention to public APIs, where usage is beyond your control and can't be predicted. Any code you release for other plugins to use should be held to the highest quality bar and tested with many permutations.

Every time a bug is encountered that wasn't caught by tests, add a test to cover it.

Include end to end functional tests, especially for areas where the logic spans global state, urls and multiple plugin APIs.

Retrospectively gauge the quality of the code you ship by tracking how many bugs are reported for features that are released. Would more tests have helped reduce this number? 

Include failure cases, edge cases, and happy paths in your tests.

Reply:
I love it! I'll incorporate your changes in the next commit.

> I don't think product, team and tech leads will feel like they can't make exceptions just because we don't have it written down here. I worry it could be abused and end up as a poor excuse for less test coverage.

I had the same thought this weekend and am glad to see us converging on the same conclusion. We aren't robots and if good cause for an exception arises, we trust individuals to press their case.

Reply:
👍  to the suggested changes and conclusions reached here.